### DISCUSSION TITLE
How can I set multiple arbitrary leader keys?

### DISCUSSION THREAD

**Discussion by jgarte** [meow-edit/meow#584] [Q&A] [answered]
Hi,

How can I set multiple arbitrary leader keys?

For example, helix uses the key `g` (QWERTY layout) as a leader key in addition to the `SPC` key.

## Leader g

![image](https://github.com/meow-edit/meow/assets/47760695/caa52c79-8b7f-4db6-8bce-5336bc95ba90)

## Leader SPC

![image](https://github.com/meow-edit/meow/assets/47760695/b720d194-b442-4bbc-b306-4a711f5c1fc5)

## Leader z

![image](https://github.com/meow-edit/meow/assets/47760695/d993d58e-fd9e-42d8-a3c1-5bdea178de0f)

**Comment by adimit** [comment, marked as answer]
My current approach is to use `transient`, and set up key bindings in meow's normal mode to call these different transients.

Here's a `transient` somewhat close to Helix's `g`:

```emacs-lisp
(transient-define-prefix dispatch-goto-menu () "This isn't documentation"
  [["Move"
    ("e" "bottom" end-of-buffer)
    ("g" "top" beginning-of-buffer)
    ("d" "definition (xref)" xref-find-definitions)
    ("h" "beginning of line" beginning-of-line)
    ("l" "end of line" end-of-line)
    ("s" "first non-blank-line" beginning-of-line-text)]
   ["Buffer"
    ("n" "next buffer" next-buffer)
    ("p" "previous buffer" previous-buffer)
    ("b" "bury buffer" bury-buffer)
    ("u" "unbury buffer" unbury-buffer)
    "Avy"
    ("c" "goto char" avy-goto-char)
    ("L" "got line" avy-goto-line)]
   ["LSP"
    ("r" "references" lsp-ui-peek-find-references)
    "Files"
    ("f" "file at point" find-file-at-point)
    ("a" "last file" pop-global-mark)]])
```

Just bind it like this:

```emacs-lisp
(meow-normal-define-key
  '("g" . dispatch-goto-menu))
```

**Reply by jgarte** [reply]
Awesome! Thanks for sharing that.

**Comment by zetashift** [comment]
I actually do it like this(also a Helix user):

```lisp
(defvar-keymap find-things-keymap
  :doc "find some things, cause I forget."
  :prefix 'file-things
  "b" #'consult-dir
  "f" #'consult-fd
  "r" #'recentf
  "s" #'save-buffer
  "t" #'treemacs
  "g" #'consult-ripgrep
  "p" #'open-emacs-init-file
  "y" #'copy-file-path
  "w" #'consult-dir)
  ```
  
  And then use it in meow like this:
  
  ```lisp
    (meow-leader-define-key
   ;; SPC j/k will run the original command in MOTION state.
   '("j" . "H-j")
   '("k" . "H-k")

   (cons "f" find-things-keymap)
   (cons "c" code-keymap)
   ...etc
   ```

**Reply by adimit** [reply]
Do you get a menu for the keybinds then, too? (Like in Helix.) Maybe via `which-key`?

**Reply by zetashift** [reply]
Yes, I don't believe it uses `which-key`, tho I still use `which-key` for non-meow keybindings, but here's what I get when I press `g` :

![image](https://github.com/meow-edit/meow/assets/1857826/de67d67d-448b-4e85-ae34-cf82b1c30afc)

**Reply by adimit** [reply]
Nice. @jgarte This would solve your problem, too. The trade-off is, that while making a new keymap is more lightweight, `transient` allows you to sort actions by topic (`which-key` just lists them alphabetically.) `which-key` typically tries to stay out of your way, too, whereas `transient` pops up immediately (like in Helix).

**Reply by willbush** [reply]
@zetashift is it easy to jump between meow and helix?

**Reply by zetashift** [reply]
@willbush depends if you are a heavy Emacs user. A lot of stuff just isn't available in helix.

It's not bad at all though.

**Comment by brabalan** [comment]
Thank you for this discussion, it was most helpful to factorize keybindings between meow-leader and plain emacs. I had to go through the transient option because the keymap one requires the keymap to be defined when bound, and it’s not necessarily the case (unless I load everything before meow).

**Comment by witek** [comment]
What if I want my custom "leader key" to show different keys depending on the current major mode?

When in org-mode, I want `f` to use my-orgmode-f-map. But when in clojure-mode, I want `f` to use my-clojure-f-map.

Any ideas how to achieve this? With a keymap or with transient?
